src/stores/NotificationsStore.js

```diff
@@ -0,0 +1,280 @@
+  action,
+  computed,
+  observable,
+  reaction,
+  runInAction,
+  when,
+} from 'mobx';
+import nullthrows from 'nullthrows';
+import FCM, { FCMEvent } from 'react-native-fcm';
+import DeviceInfo from 'react-native-device-info';
+import AuthStore from './AuthStore';
+import { AchievementStore, KegStore } from './DAOStores';
+import { fetchJSON } from './ApiRequestStores/makeRequestApiStore';
+import CONFIG from '../config';
+
+const BASE_PUSH_URL = `${CONFIG.HOST}api/v2/push`;
+
+const NOTIFICATIONS_STORAGE_KEY = 'notifications';
+
+export type BaseNotification<TNotificationExtraProps: {}> = {
+  ...TNotificationExtraProps,
+  body: string,
+  date: Date,
+  id: string,
+  isRead: string,
+  title: string,
+};
+
+export type LowKegLevelNotificationExtraProps = {
+  beverageId: EntityID,
+  beverageName: string,
+  tapId: EntityID,
+  type: 'lowKegLevel',
+};
+
+export type NewAchievementNotificationExtraProps = {
+  achievementType: AchievementType,
+  type: 'newAchievement',
+};
+
+export type NewFriendRequestNotificationExtraProps = {
+  friendId: EntityID,
+  type: 'newFriendRequest',
+  friendUserName: string,
+};
+
+export type LowKegLevelNotification = BaseNotification<
+  LowKegLevelNotificationExtraProps,
+>;
+
+export type NewAchievementNotification = BaseNotification<
+  NewAchievementNotificationExtraProps,
+>;
+
+export type NewFriendRequestNotification = BaseNotification<
+  NewFriendRequestNotificationExtraProps,
+>;
+
+export type Notification =
+  | LowKegLevelNotification
+  | NewAchievementNotification
+  | NewFriendRequestNotification;
+
+class NotificationsStore {
+  @observable _navigation: ?Navigation = null;
+  @observable
+  _notificationsByID: ObservableMap<BaseNotification<any>> = observable.map();
+
+  constructor() {
+    AppState.addEventListener('change', nextAppState => {
+      if (nextAppState === 'active') {
+        FCM.removeAllDeliveredNotifications();
+      }
+    });
+
+    FCM.on(FCMEvent.Notification, this._onRawNotification);
+
+    (async () => {
+      await this._rehydrateNotifications();
+
+      reaction(
+        () => this._notificationsByID.values(),
+        notifications => {
+          AsyncStorage.setItem(
+            NOTIFICATIONS_STORAGE_KEY,
+            JSON.stringify(notifications),
+          );
+        },
+        {
+          compareStructural: true,
+        },
+      );
+    })();
+  }
+
+  @computed
+  get notifications(): Array<Notification> {
+    return this._notificationsByID
+      .values()
+      .sort((a, b): number => new Date(b.date) - new Date(a.date));
+  }
+
+  @computed
+  get unreadCount(): number {
+    return this._notificationsByID
+      .values()
+      .filter(({ isRead }: Notifications) => !isRead).length;
+  }
+
+  @computed
+  get hasUnread(): boolean {
+    return this.unreadCount > 0;
+  }
+
+  @computed
+  get _isInitialized(): boolean {
+    return !!this._navigation && AuthStore.isAuthorized;
+  }
+
+  handleInitialNotification = async () => {
```

Comment:
How do you handle the case where I got 5 notifications without the app open. I open the app without clicking one of those notifications? If it's covered somewhere then I missed it.

Reply:
If 5 notifications came and you open the app, all notifications are on the notifications screen and all removed from the Android tray. If you open the app by clicking on one of those notifications it redirects to what this notif routes to. The other logic is the same as in the 1st case.

Reply:
Sweet. That's exactly what we want.